File: find_duplicate.rb
def find_duplicate(nums)
  one = nums.first
  two = nums.first
  while true
    one = nums[one]
    two = nums[nums[two]]
    break if one == two

  # p "#{one} one"
  # p "#{two} two"
  end

  ptr1 = nums.first
  ptr2 = one
  while ptr1 != ptr2
    ptr1 = nums[ptr1]
    ptr2 = nums[ptr2]
  end
  ptr1
end

# O(n)
# def find_duplicate(nums)
#   hash = {}
#   nums.each do |el|
#     return el if hash[el]
#     hash[el] = true
#   end
# end

arr = [1,3,3,2]

p find_duplicate(arr)

File: findDuplicate.js
var findDuplicate = function(nums) {
  let hash = {};
  for (let el of nums) {
    if (hash[el]) {
      return el;
    }
    hash[el] = true;
  }
};

// var findDuplicate = function(nums) {
//   let tortoise = nums[0];
//   let hare = nums[0];
//   while (true) {
//     tortoise = nums[tortoise];
//     hare = nums[nums[hare]];
//     if (hare === tortoise) {
//       break;
//     }
//   }
//
//   let ptr1 = tortoise;
//   let ptr2 = nums[0];
//   while (ptr1 !== ptr2) {
//     ptr1 = nums[ptr1];
//     ptr2 = nums[ptr2];
//   }
//
//   return ptr1;
// };

let arr = [2,5,9,6,9,3,8,9,7,1];

console.log(findDuplicate(arr));
